src/librustc_privacy/lib.rs

```diff
@@ -20,22 +20,260 @@ use rustc::hir::intravisit::{self, Visitor, NestedVisitorMap};
+                        Node::Item(item) => match &item.node {
+                            hir::ItemKind::Impl(.., None, _, _) => &impl_item.vis,
+                            hir::ItemKind::Impl(.., Some(trait_ref), _, _)
+                                => return def_id_visibility(tcx, trait_ref.path.def.def_id()),
+                            kind => bug!("unexpected item kind: {:?}", kind),
+                        }
+                        node => bug!("unexpected node kind: {:?}", node),
+                    }
+                }
+                Node::StructCtor(vdata) => {
+                    let struct_node_id = tcx.hir().get_parent(node_id);
+                    let item = match tcx.hir().get(struct_node_id) {
+                        Node::Item(item) => item,
+                        node => bug!("unexpected node kind: {:?}", node),
+                    };
+                    let (mut ctor_vis, mut span, mut descr) =
+                        (ty::Visibility::from_hir(&item.vis, struct_node_id, tcx),
+                         item.vis.span, item.vis.node.descr());
+                    for field in vdata.fields() {
+                        let field_vis = ty::Visibility::from_hir(&field.vis, node_id, tcx);
+                        if ctor_vis.is_at_least(field_vis, tcx) {
+                            ctor_vis = field_vis;
+                            span = field.vis.span;
+                            descr = field.vis.node.descr();
+                        }
+                    }
+
+                    // If the structure is marked as non_exhaustive then lower the
+                    // visibility to within the crate.
+                    if ctor_vis == ty::Visibility::Public {
+                        let adt_def = tcx.adt_def(tcx.hir().get_parent_did(node_id));
+                        if adt_def.non_enum_variant().is_field_list_non_exhaustive() {
+                            ctor_vis = ty::Visibility::Restricted(DefId::local(CRATE_DEF_INDEX));
+                            span = attr::find_by_name(&item.attrs, "non_exhaustive").unwrap().span;
+                            descr = "crate-visible";
+                        }
+                    }
+
+                    return (ctor_vis, span, descr);
+                }
+                Node::Expr(expr) => {
+                    return (ty::Visibility::Restricted(tcx.hir().get_module_parent(expr.id)),
+                            expr.span, "private")
+                }
+                node => bug!("unexpected node kind: {:?}", node)
+            };
+            (ty::Visibility::from_hir(vis, node_id, tcx), vis.span, vis.node.descr())
+        }
+        None => (tcx.visibility(def_id), tcx.def_span(def_id), "private"),
+    }
+}
+
+// Set the correct `TypeckTables` for the given `item_id` (or an empty table if
+// there is no `TypeckTables` for the item).
+fn update_tables<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
```

Comment:
style: I would out-of-line the `mem::replace` - have this function return the right `&'a ty::TypeckTables<'tcx>`, and have the call-sites do `mem::replace(&mut self.tables, get_tables(tcx, item_id, empty_tables))`.

Reply:
pre-existing, I know.

Reply:
also, style: I think the normal style is to use the function as `mem::replace` and not as `replace`.